psycopg/psycopgmodule.c

```diff
@@ -109,12 +189,16 @@ psyco_connect(PyObject *self, PyObject *args, PyObject *keywds)
         conn = PyObject_CallFunction(factory, "si", dsn, async);
```

Comment:
Sorry, I can't just stop :-p

Using `PQconnectdbParams()` would also allow us to lift the limitation of either dsn or keyword arguments that exists in current `psycopg2.connect()` implementation, because this function is able to do just that: take a dsn in either conninfo or URI form and apply any number of keyword arguments on top of it, to produce the actual set of connection parameters.

There is a problem with this approach, however: extracting the exact dsn string *that was passed to connect function* is no longer possible, because there is no such single string (well, we can still combine `parse_dsn` and `make_dsn` in creative ways...)

What is possible, is extracting the dsn *that was used to establish the connection*, using `PQconninfo()`, but it's only available from 9.3 and it is going to add a number of default values such as `client_encoding=utf8`.  What it will also do is reveal any connection parameters that are set from environment variables, which might be a useful thing in itself.  I think we should add support for `PQconninfo()` separately, regardless of the status of this patch.